### PR TITLE
Search implementation

### LINKED ISSUE #30: Search
Needs https://github.com/FUMR/music-service-async-interface/issues/14 to be resolved.

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
I need some feedback guys ;)

ATM It depends on commit from draft PR, will be changed later.

depends on FUMR/music-service-async-interface#31
resolves #30

### CODE REVIEW

**Review by divadsn — APPROVED** [review]
The search implementation itself is very identical to the 0.7.x version of https://github.com/tamland/python-tidal, LGTM.

File: tidal_async/api.py [hunk 1]
```diff
@@ -243,28 +247,21 @@ def cover(self) -> Optional[Cover]:
...
-    def artist(self) -> "Artist":
-        """
-        :return: :class:`Track`'s :class:`Artist`
-        """
-        return Artist(self.sess, self["artist"])
-
-    async def artists(self) -> AsyncGenerator[Tuple["Artist", ArtistType], None]:
-        """Generates async interable of :class:`Artist`s of the :class:`Track` with their corresponding role
+    def artists(self) -> List[Tuple["Artist", ArtistType]]:
+        """Generates :class:`list` of :class:`Artist`s of the :class:`Track` with their corresponding role
```
Comment (wvffle) [inline tidal_async/api.py:251]: Compared to L262, you have to decide whether you use `list` or `:class:`list``

File: tidal_async/api.py [hunk 2]
```diff
@@ -479,41 +468,42 @@ async def tracks(self, per_request_limit: int = 50) -> AsyncGenerator[Track, Non
...
         """
         :return: :class:`Cover` image or `None` if :class:`Album` has no cover
         """
         return Cover(self.sess, self["cover"]) if self["cover"] is not None else None
 
-    async def artists(self) -> AsyncGenerator[Tuple["Artist", ArtistType], None]:
-        """Generates async interable of :class:`Artist`s of the :class:`Album` with their corresponding role
+    @property
+    def artists(self) -> List[Tuple["Artist", ArtistType]]:
+        """Generates :class:`list` of :class:`Artist`s of the :class:`Album` with their corresponding role
```
Comment (wvffle) [inline tidal_async/api.py:495]: Same here

File: tidal_async/session.py [hunk 3]
```diff
@@ -282,6 +282,67 @@ def is_valid_url(url: str) -> bool:
...
+        [<tidal_async.api.Track (150607332): Wiatr, Sobel, be vis - Testarossa>,
+         <tidal_async.api.Track (96312544): Joji - TEST DRIVE>,
+         <tidal_async.api.Artist (14813): Testament>,
+         <tidal_async.api.Artist (25733): Crash Test Dummies>]
+        >>> [i async for i in sess.search("test", Track, 2)]
+        [<tidal_async.api.Track (150607332): Wiatr, Sobel, be vis - Testarossa>,
+         <tidal_async.api.Track (96312544): Joji - TEST DRIVE>]
+
+        :param query: search string/phrase
+        :param types: :class:`Searchable` types or list of types, filters search results to this type,
```
Comment (wvffle) [inline tidal_async/session.py:315]: One can `filter things out` and `restrict things to`

File: tidal_async/session.py [hunk 4]
```diff
@@ -282,6 +282,67 @@ def is_valid_url(url: str) -> bool:
...
+        e.g. if you want to search for tracks use `sess.search('text', Track)`
+        or if you want to search for albums and playlists use `sess.search('text', [Album, Playlist])`.
+        Also allows to search for all possible objects, then this parameter is left unset or set to None/empty list.
```
Comment (wvffle) [inline tidal_async/session.py:318]: s#is left unset or set to None/empty list.#is unset, set to None or an empty list.#

File: tidal_async/session.py [hunk 5]
```diff
@@ -282,6 +282,67 @@ def is_valid_url(url: str) -> bool:
...
+        :param limit: amount of records per type to search for
+        Maximum amount of results will be `len(types) * limit`.
+        :raises InvalidSearchType: when `types` contains invalid type
+        (not subclass of :class:`Searchable` or :class:`TidalObject`)
```
Comment (wvffle) [inline tidal_async/session.py:322]: s/not/not a/

File: tidal_async/session.py [hunk 6]
```diff
@@ -282,6 +282,67 @@ def is_valid_url(url: str) -> bool:
...
+        [<tidal_async.api.Track (150607332): Wiatr, Sobel, be vis - Testarossa>,
+         <tidal_async.api.Track (96312544): Joji - TEST DRIVE>]
+
+        :param query: search string/phrase
+        :param types: :class:`Searchable` types or list of types, filters search results to this type,
+        e.g. if you want to search for tracks use `sess.search('text', Track)`
+        or if you want to search for albums and playlists use `sess.search('text', [Album, Playlist])`.
+        Also allows to search for all possible objects, then this parameter is left unset or set to None/empty list.
+        :param limit: amount of records per type to search for
+        Maximum amount of results will be `len(types) * limit`.
```
Comment (wvffle) [inline tidal_async/session.py:320]: is it how the Tidal API works?

**Review by wvffle — CHANGES_REQUESTED** [review]

File: tidal_async/session.py (hunk 6 above)
Comment (JuniorJPDJ) [reply tidal_async/session.py:320]: yop

Comment (wvffle) [reply tidal_async/session.py:320]: hmm, kind of weird

File: tidal_async/api.py (hunk 1 above)
Comment (JuniorJPDJ) [reply tidal_async/api.py:251]: Well. No.

After I introduce type here, I don't use type spec in `:return:` as it's not needed.
I'll change it if you are complaining tho xD

File: tidal_async/session.py (hunk 3 above)
Comment (JuniorJPDJ) [reply tidal_async/session.py:315]: It's not the same but your version is fitting better.

File: tidal_async/session.py (hunk 4 above)
Comment (JuniorJPDJ) [reply tidal_async/session.py:318]: I don't like your version.
What do you think about:
then this parameter is left unset, set to None or empty list.

Comment (wvffle) [reply tidal_async/session.py:318]: or an

File: tidal_async/session.py (hunk 6 above)
Comment (JuniorJPDJ) [reply tidal_async/session.py:320]: true, but well.. I can do nothing about it as I'd like not to divide offset

**Comment by JuniorJPDJ** [comment]
> The search implementation itself is very identical to the 0.7.x version of [tamland/python-tidal](https://github.com/tamland/python-tidal), LGTM.

It's totally other than python-tidal, which is closer to how API itself works.
I wanted to do this in generic way so we can use search in similar ways for every other service.
The difference is in a way it returns values.

Anyway - thanks for positive feedback @divadsn :D

**Review by wvffle — APPROVED** [review]

**Comment by JuniorJPDJ** [comment]
I just changed dep commit to point on merged FUMR/music-service-async-interface#31.
I think it's ready to go.
I'm now playing with tests and CI on #76 so I'll fix that caching bug there - tests will not pass on this PR.

**Pull request merged**